components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java

```diff
@@ -170,32 +180,69 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
                 + ":" +
                 validationContext.getHttpVerb();
 
-        boolean isValid = false;
+        Set<OAuth2ScopeValidator> oAuth2ScopeValidators = OAuthServerConfiguration.getInstance()
+                .getOAuth2ScopeValidators();
+        //validate scope for filtered validators from db
+        String[] scopeValidators;
+        OAuthAppDO appInfo;
+        OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
+
         try {
-            Set<OAuth2ScopeValidator> oAuth2ScopeValidators = OAuthServerConfiguration.getInstance()
-                    .getOAuth2ScopeValidators();
-            //validate scope from all the validators
+            appInfo = oAuthAppDAO.getAppInformation(clientId);
+            scopeValidators = appInfo.getScopeValidators();     //get scope validators from the DB
+            List<String> appScopeValidators=  new ArrayList<>(Arrays.asList(scopeValidators));
+
+            if (ArrayUtils.isEmpty(scopeValidators)) {
+                log.debug(String.format("There is no scope validator registered for %s@%s", appInfo.getApplicationName(),
+                            OAuth2Util.getTenantDomainOfOauthApp(appInfo)));
+                return true;
+            }
+
             for (OAuth2ScopeValidator validator : oAuth2ScopeValidators) {
-                if (validator != null) {
-                    if (validator.validateScope(accessTokenDO, resource)) {
-                        isValid = true;
-                    } else {
-                        if (log.isDebugEnabled()) {
-                            log.debug("Scope validation failed from "+ validator);
+                try {
+                    if (validator != null && appScopeValidators.contains(validator.getValidatorName())) {
+                        //take the intersection of defined scope validators and scope
+                        //validators registered for the application
+                        log.debug(String.format("Validating scope of token %s using %s", accessTokenDO.getTokenId(),
+                                validator.getValidatorName()));
+
+                        boolean isValid = validator.validateScope(accessTokenDO, resource);
+                        if (!isValid) {
+                            if (log.isDebugEnabled()) {
```

Comment:
Why are we following two approaches for debug logs?
checking for `log.isDebugEnabled()` and not checking `log.isDebugEnabled()`

Reply:
Here we want only one approach to print the message when the condition is like scope validation fails and log.isDebugEnabled() = true

Reply:
Resolved